### PR TITLE
feat: th0th improvements — Synapse layer, search quality, enriched mode, MCP completeness

### PR DESCRIPTION
**Pull request by S1LV4**
## Summary

This PR consolidates all improvements developed this sprint:

### 🧠 Synapse — cognitive modulation layer (new feature)
Post-retrieval intelligence pipeline with 8 configurable stages: buffer merge, pre-filter, attention score re-ranking, chain inhibition (intent-based boost/suppress), diversity penalty (MMR dedup), temporal inhibition, confidence gate, score spectrum. Full session management API (`/api/v1/synapse/session/*`), 25 service files, 20+ test files.

### 🐛 Bug fixes
- **`keyword-search-pg`**: code tokens with `(`, `:`, `"` broke `to_tsquery` silently, zeroing all search results
- **`memory-repository`**: punctuation in FTS5 query caused `MATCH ''` errors; now strips special chars + graceful fallback
- **`ignore-patterns`**: missing `**/` prefix meant `node_modules/` only matched at tree root — monorepos had `packages/*/node_modules` indexed
- **Keyword search failure**: error in BM25 no longer kills the whole `Promise.all` — falls back to `[]`
- **`search-cache`**: robust SQLite DB init with retry for fresh installs

### 🔍 Search quality (all env-configurable, defaults backward-compatible)
- `SEARCH_DISABLE_KEYWORD=true` — pure vector-only (**+44% MRR** measured on CoIR cosqa)
- `RRF_KEYWORD_BOOST`, `RRF_VECTOR_WEIGHT`, `RRF_MAX_CHUNKS_PER_FILE` — tuning knobs
- `SEARCH_MIN_SCORE`, `OLLAMA_EMBED_DELAY_MS` — calibration knobs
- Indexing `BATCH_SIZE` 10→20

### 📦 `responseMode="enriched"` (new)
Returns full content + `fileImports` + `parentSymbol` + `chunkIndex`/`totalChunks` in one call — eliminates most follow-up Read/grep calls. smart-chunker now pre-computes these at index time.

### 🔌 MCP client — 7 new tools (14 → 21)
`th0th_read_file`, `th0th_synapse_session`, `th0th_synapse_prime`, `th0th_synapse_access`, `th0th_symbol_snippet`, `th0th_memory_list`, `th0th_reindex`

### 📚 SKILL.md v3
Importance rubric (5-question deterministic scoring), memory tiers, session naming conventions, debug recall-first pattern, evidence gates for completion.

---

## Migration notes
- **Re-index required** after merging to populate `fileImports`/`parentSymbol` in chunks:
  ```
  th0th_reset_project({ projectId: "your-project" })
  th0th_index({ projectPath: "/path/to/project", projectId: "your-project" })
  ```

## Test plan
- [ ] `bun test` — 626+ pass, 1 pre-existing fail (MemoryConsolidationJob integration)
- [ ] TypeScript clean across all workspaces
- [ ] Search with punctuation in query (`?`, `!`) no longer zeros memory results
- [ ] `packages/*/node_modules` no longer appears in monorepo index
- [ ] `responseMode="enriched"` returns `fileImports` and `parentSymbol` (post re-index)
- [ ] `th0th_synapse_session` creates session and returns `sessionId`
- [ ] `th0th_memory_list` lists memories without semantic query

## Summary by CodeRabbit

* **New Features**
  * Full Synapse cognitive layer (sessions, working-memory buffer, attention re-ranking, diversity/temporal/intent/confidence gates, priming/prefetch, embedding evolution, selective compression/strengthening) and new enriched search mode returning content plus file/import context and chunk navigation.

* **Improvements**
  * Type-scoped memory search, more resilient cache/DB handling, safer tokenization/FTS behavior, configurable embedding request delay, improved preview generation, and refined ignore patterns.

* **Documentation**
  * Synapse usage guide, updated tool/memory docs, and benchmark analysis/run scripts.

* **Tests**
  * Large suite of regression and IR tests added to improve reliability.

### CODE REVIEW

File: apps/mcp-client/src/tool-definitions.ts [hunk 1]
```diff
@@ -517,6 +517,118 @@ export const TOOL_DEFINITIONS: ToolDefinition[] = [
...
+        compress: { type: "boolean", description: "Auto-compress content > 100 lines (default: true)", default: true },
+        includeSymbols: { type: "boolean", description: "Include symbol definitions/references (default: true)", default: true },
+        includeImports: { type: "boolean", description: "Extract file imports (default: true)", default: true },
+      },
+      required: ["filePath"],
+    },
+  },
+  {
+    name: "th0th_synapse_session",
+    description: "Create/resume a Synapse cognitive session. Returns sessionId to pass as synapseSessionId on every th0th_search. Activates task alignment, agent affinity, working-memory buffer. Name by intent: 'debug-auth', 'feature-payment'.",
```
[the remark this answers is not in the document]
Comment (S1LV4) [reply apps/mcp-client/src/tool-definitions.ts]: Fixed in 67ea4ca — updated description to use `sessionId` instead of `synapseSessionId`.

File: packages/core/src/__tests__/synapse-edge-cases.test.ts [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    const inputs: SearchResult[] = [];
+    for (let i = 0; i < 100; i++) {
+      inputs.push(r(`x${i}`, 1 - i * 0.005, { filePath: `src/file-${i % 10}.ts` }));
+    }
+    const start = performance.now();
+    const out = applyAttentionScore(inputs, { ...DEFAULT_ATTENTION_CONFIG, enabled: true, rerankWindow: 50 }, null);
+    const elapsed = performance.now() - start;
+    expect(out.results).toHaveLength(100);
+    expect(elapsed).toBeLessThan(50); // <50ms for 100 candidates
+  });
```
[the remark this answers is not in the document]
Comment (S1LV4) [reply packages/core/src/__tests__/synapse-edge-cases.test.ts:212]: Fixed in 67ea4ca — replaced hard `50` with `Number(process.env.SYNAPSE_ATTENTION_MAX_MS ?? "250")`.

File: packages/core/src/services/search/contextual-search-rlm.ts [hunk 3]
```diff
@@ -723,7 +737,8 @@ export class ContextualSearchRLM {
     // Keyword weight multiplier (higher = more weight to keyword results)
     // For code queries: 2.5x boost to keyword matches
     // For general queries: 1.0x (equal weight)
-    const KEYWORD_BOOST = isCodeQuery ? 2.5 : 1.0;
+    const codeKeywordBoost = Number(process.env.RRF_KEYWORD_BOOST ?? "2.5");
+    const KEYWORD_BOOST = isCodeQuery ? codeKeywordBoost : 1.0;
```
[the remark this answers is not in the document]
Comment (S1LV4) [reply packages/core/src/services/search/contextual-search-rlm.ts]: Fixed in 67ea4ca — added NaN/range validation for both `RRF_KEYWORD_BOOST` and `RRF_VECTOR_WEIGHT` before score math.

File: packages/core/src/services/synapse/buffer/working-memory-buffer.ts [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+        for (const t of queryTokens) existing.queryTokens.add(t);
+        existing.lastAccessedAt = now;
+        // Only overwrite baselineScore when an explicit raw score was given.
+        // Otherwise the original baseline survives — preventing drift.
+        if (baseline != null) existing.baselineScore = baseline;
+      } else {
+        this.entries.set(result.id, {
+          result,
+          queryTokens: new Set(queryTokens),
+          contentTokens: tokenize(result.content || ""),
```
[the remark this answers is not in the document]
Comment (S1LV4) [reply packages/core/src/services/synapse/buffer/working-memory-buffer.ts:161]: Fixed in 19e98c4 — `put()` now refreshes `existing.result` and `existing.contentTokens` on re-insert.

File: packages/core/src/services/synapse/scoring/attention-score.ts [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    return { results, breakdowns: [] };
+  }
+
+  // IMP-14: cap the rerank window by the number of results we actually
+  // intend to surface (`config.rerankWindow`). When the caller passes a
+  // huge ceiling but only N results are available, doing the work for
+  // anything beyond the displayed window is wasted. The lower bound of
+  // 10 guarantees enough breathing room for ties/diversity at the tail.
+  const effective = Math.max(10, config.rerankWindow);
+  const window = Math.min(effective, results.length);
```
[the remark this answers is not in the document]
Comment (S1LV4) [reply packages/core/src/services/synapse/scoring/attention-score.ts]: Fixed in 19e98c4 — replaced `Math.max(10, ...)` with `config.rerankWindow > 0 ? Math.floor(config.rerankWindow) : 10` so positive values below 10 are honored while 0 still falls back safely.

File: packages/core/src/services/synapse/session/session-registry.ts [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+      taskEmbedding: input.taskEmbedding,
+      createdAt: now,
+      expiresAt: now + ttl,
+      accessHistory: new Map(),
+      accessHistoryLimit: input.accessHistoryMaxEntries ?? DEFAULT_ACCESS_HISTORY_LIMIT,
+      buffer: input.bufferConfig ? new WorkingMemoryBuffer(input.bufferConfig) : undefined,
+    };
+    this.sessions.set(session.sessionId, session);
+    return session;
+  }
```
[the remark this answers is not in the document]
Comment (S1LV4) [reply packages/core/src/services/synapse/session/session-registry.ts:73]: Fixed in 19e98c4 — `create()` now throws if the sessionId already exists.

File: packages/core/src/services/synapse/session/session-registry.ts [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    const session = this.sessions.get(sessionId);
+    if (!session) return null;
+    if (session.expiresAt <= now) {
+      this.sessions.delete(sessionId);
+      return null;
+    }
+    const refreshed = now + this.defaultTtlMs;
+    if (refreshed > session.expiresAt) {
+      session.expiresAt = refreshed;
+    }
```
[the remark this answers is not in the document]
Comment (S1LV4) [reply packages/core/src/services/synapse/session/session-registry.ts]: Fixed in 19e98c4 — stored `ttlMs` on `AgentSession`; `get()` and `updateTaskContext()` now use `session.ttlMs` instead of `this.defaultTtlMs`.

File: packages/core/src/services/synapse/synapse-manager.ts [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+      stream = out.results;
+      applied.push("attention");
+    }
+
+    const chainCfg = this.config.inhibition.chainInhibition;
+    if (chainCfg?.enabled) {
+      const out = applyChainInhibition(stream, query, {
+        enabled: true,
+        boosts: DEFAULT_CHAIN_BOOSTS,
+      });
```
[the remark this answers is not in the document]
Comment (S1LV4) [reply packages/core/src/services/synapse/synapse-manager.ts:143]: Fixed in 19e98c4 — passes `chainCfg.boosts ?? DEFAULT_CHAIN_BOOSTS` so runtime config is respected.

File: packages/shared/src/config/index.ts [hunk 9]
```diff
@@ -253,6 +369,34 @@ export class Config {
       rateLimit: { ...defaults.rateLimit, ...overrides.rateLimit },
       security: { ...defaults.security, ...overrides.security },
       logging: { ...defaults.logging, ...overrides.logging },
+      synapse: {
+        ...defaults.synapse,
+        ...overrides.synapse,
+        inhibition: {
+          ...defaults.synapse.inhibition,
+          ...overrides.synapse?.inhibition,
+        },
```
[the remark this answers is not in the document]
Comment (S1LV4) [reply packages/shared/src/config/index.ts:399]: Fixed in 19e98c4 — deep-merged `diversityPenalty`, `temporalInhibition`, `confidenceGate` (including `thresholds`), and `chainInhibition` sub-objects.

File: apps/tools-api/src/routes/memory.ts [hunk 10]
```diff
@@ -158,53 +158,37 @@ export const memoryRoutes = new Elysia({ prefix: "/api/v1/memory" })
...
+        } else {
+          // PostgreSQL path — search method
+          const all = await repo.search({
+            types: body.type ? [body.type] : undefined,
+            minImportance: body.minImportance ?? 0,
+            includePersistent: true,
+            limit: 10000,
+          });
+          total = all.length;
+          rows = all.slice(offset, offset + limit);
```
[the remark this answers is not in the document]
Comment (S1LV4) [reply apps/tools-api/src/routes/memory.ts:186]: Acknowledged. Requires a `count()` method on the repository layer — tracked as follow-up, out of scope for this PR.

File: packages/core/src/scripts/symbol-benchmark.ts [hunk 11]
```diff
@@ -181,7 +181,7 @@ function measure<T>(fn: () => Promise<T>): Promise<{ result: T; ms: number }> {
 const DEFINITION_FIXTURES: Array<{ name: string; kind?: string; minRefs: number }> = [
   { name: "ContextualSearchRLM", kind: "class", minRefs: 2 },
   { name: "symbolGraphService", kind: "variable", minRefs: 1 },
-  { name: "contextualSearch", kind: "variable", minRefs: 1 },
+  { name: "ContextualSearchRLM", kind: "class", minRefs: 1 },
```
[the remark this answers is not in the document]
Comment (S1LV4) [reply packages/core/src/scripts/symbol-benchmark.ts]: Fixed in 43ab2e3 — removed the redundant `minRefs: 1` entry, keeping only the original `minRefs: 2` fixture.
